data.go

```diff
@@ -2,104 +2,136 @@ package main
-		return nil
+func (s *server) setIconInto(img *canvas.Image) {
+	s.avatarMu.Lock()
+	avatar := s.avatar
+	s.avatarMu.Unlock()
+
+	if avatar != nil {
+		img.Resource = avatar
 	}
 
-	return icon
+	s.avatarOnce.Do(func() {
+		go func() {
+			// TODO cache this resource
+			icon, err := fyne.LoadResourceFromURLString(s.iconURL)
+			if err != nil {
+				fyne.LogError("Failed to read icon "+s.iconURL, err)
+				return
+			}
+
+			s.avatarMu.Lock()
+			s.avatar = icon
+			s.avatarMu.Unlock()
+
+			img.Refresh()
+		}()
+	})
 }
 
 type channel struct {
-	id       int
-	name     string
-	messages []*message
+	id   uint64
+	name string
 }
 
 type message struct {
-	content, author string
-	avatar          string
+	content string
+	author  string
+	avatar  string
 }
 
 func (u *ui) loadChannels() {
 	for _, s := range u.data.servers {
-		cs, _ := u.conn.Client.Channels(discord.GuildID(s.id))
+		cs, _ := u.conn.Channels(discord.GuildID(s.id))
 		for _, c := range cs {
-			if c.Type == discord.GuildCategory || c.Type == discord.GuildVoice {
+			if c.Type != discord.GuildText {
 				continue // ignore voice and groupings for now
 			}
 
-			chn := &channel{id: int(c.ID), name: c.Name}
-			if len(s.channels) == 0 {
-				chn.messages = u.loadRecentMessages(c.ID)
-				if s == u.currentServer {
-					u.currentChannel = chn
-					u.messages.Objects = nil
-					u.appendMessages(u.currentChannel.messages)
-				}
-			}
+			chn := &channel{id: uint64(c.ID), name: c.Name}
 			s.channels = append(s.channels, chn)
 		}
 	}
 	u.channels.Refresh()
+}
 
-	for _, s := range u.data.servers {
-		for i, c := range s.channels {
-			if i == 0 {
-				continue // we did this one above
-			}
-			c.messages = u.loadRecentMessages(discord.ChannelID(c.id))
-		}
-	}
+func (u *ui) renderMessage(m *discord.Message) *message {
+	node := md.ParseWithMessage([]byte(m.Content), u.conn.Cabinet, m, true)
+	var builder strings.Builder
+	md.DefaultRenderer.Render(&builder, []byte(m.Content), node)
+
+	return &message{author: m.Author.Username,
+		content: strings.TrimSuffix(builder.String(), "\n"),
+		avatar:  m.Author.AvatarURLWithType(discord.PNGImage) + "?size=64"}
 }
 
 func (u *ui) loadRecentMessages(id discord.ChannelID) []*message {
-	ms, err := u.conn.Client.Messages(id, 15)
+	ms, err := u.conn.Messages(id)
 	if err != nil {
 		return nil
 	}
 
 	var list []*message
+	var subscribed bool
+
 	for i := len(ms) - 1; i >= 0; i-- { // newest message is first in response
 		m := ms[i]
-		msg := &message{author: m.Author.Username, content: m.Content,
-			avatar: m.Author.AvatarURL()}
-		list = append(list, msg)
+		list = append(list, u.renderMessage(&m))
+
+		if !subscribed && m.GuildID.IsValid() {
+			u.conn.MemberState.Subscribe(m.GuildID)
+			subscribed = true
+		}
 	}
 
 	return list
 }
 
-func loadServers(s *session.Session, u *ui) {
-	var servers []*server
-	gs, err := s.Client.Guilds(0)
+func loadServers(session *session.Session, u *ui) {
```

Comment:
Not sure why this was renamed - isn't `session *session.Session` rather against Go's naming approach?